codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java

```diff
@@ -405,67 +476,54 @@ private ChannelFuture writeHeadersInternal(ChannelHandlerContext ctx, ChannelPro
-     * number of CONTINUATION frames.
-     */
-    private ChannelFuture continueHeaders(ChannelHandlerContext ctx, ChannelPromise promise,
-            int streamId, int padding, ByteBuf headerBlock, ByteBuf firstFrame) {
-        // Create a composite buffer wrapping the first frame and any continuation frames.
-        CompositeByteBuf out = ctx.alloc().compositeBuffer();
-        out.addComponent(firstFrame);
-        int numBytes = firstFrame.readableBytes();
-
-        // Process any continuation frames there might be.
-        while (headerBlock.isReadable()) {
-            ByteBuf frame = createContinuationFrame(ctx, streamId, headerBlock, padding);
-            out.addComponent(frame);
-            numBytes += frame.readableBytes();
-        }
-
-        out.writerIndex(numBytes);
-        return ctx.write(out, promise);
-    }
-
-    /**
-     * Allocates a new buffer and writes a single continuation frame with a fragment of the header
-     * block to the output buffer.
+     * Writes as many continuation frames as needed until {@code padding} and {@code headerBlock} are consumed.
      */
-    private ByteBuf createContinuationFrame(ChannelHandlerContext ctx, int streamId,
-            ByteBuf headerBlock, int padding) {
+    private ChannelFuture writeContinuationFrames(ChannelHandlerContext ctx, int streamId,
+            ByteBuf headerBlock, int padding, ChannelPromise promise) {
         Http2Flags flags = new Http2Flags().paddingPresent(padding > 0);
         int nonFragmentLength = padding + flags.getPaddingPresenceFieldLength();
         int maxFragmentLength = maxFrameSize - nonFragmentLength;
-        ByteBuf fragment =
-                headerBlock.readSlice(Math.min(headerBlock.readableBytes(), maxFragmentLength));
+        // TODO: same padding is applied to all frames, is this desired?
+        if (maxFragmentLength <= 0) {
+            return promise.setFailure(new IllegalArgumentException(
+                    "Padding [" + padding + "] is too large for max frame size [" + maxFrameSize + "]"));
+        }
+        SimpleChannelPromiseAggregator promiseAggregator = new SimpleChannelPromiseAggregator(ctx, promise);
+        while (headerBlock.isReadable()) {
+            ByteBuf fragment =
+                    headerBlock.readSlice(Math.min(headerBlock.readableBytes(), maxFragmentLength)).retain();
 
-        int payloadLength = fragment.readableBytes() + nonFragmentLength;
-        ByteBuf frame = ctx.alloc().buffer(FRAME_HEADER_LENGTH + payloadLength);
-        flags = flags.endOfHeaders(headerBlock.readableBytes() == 0);
+            int payloadLength = fragment.readableBytes() + nonFragmentLength;
+            flags = flags.endOfHeaders(!headerBlock.isReadable());
 
-        writeFrameHeader(frame, payloadLength, CONTINUATION, flags, streamId);
+            writeFrameHeader(ctx, payloadLength, CONTINUATION, flags, streamId, promiseAggregator.newPromise());
```

Comment:
They may change but only for the last continuation in the sequence so we could reuse the frame header for the first N-1 continuations.